File: dsa_leetcode_beginner/1143-longest-common-subsequence/1143-longest-common-subsequence.js
class Solution {
    longestCommonSubsequence(text1, text2) {
        const m = text1.length;
        const n = text2.length;
        const dp = Array.from({ length: m + 1 }, () => Array(n + 1).fill(0));

        for (let j = 1; j <= n; j++) {
            for (let i = 1; i <= m; i++) {
                if (text1[i - 1] === text2[j - 1]) {
                    dp[i][j] = dp[i - 1][j - 1] + 1;
                } else {
                    dp[i][j] = Math.max(dp[i][j - 1], dp[i - 1][j]);
                }
            }
        }

        // The bottom-right corner is the length of the LCS
        return dp[m][n];
    }
}

File: dsa_leetcode_beginner/1143-longest-common-subsequence/1143-longest-common-subsequence.cpp
#include <vector>
#include <string>
#include <algorithm>

class Solution {
public:
    int longestCommonSubsequence(std::string text1, std::string text2) {
        int m = text1.size();
        int n = text2.size();
        std::vector<std::vector<int>> dp(m + 1, std::vector<int>(n + 1, 0));

        for (int j = 1; j <= n; ++j) {
            for (int i = 1; i <= m; ++i) {
                if (text1[i - 1] == text2[j - 1]) {
                    dp[i][j] = dp[i - 1][j - 1] + 1;
                } else {
                    dp[i][j] = std::max(dp[i][j - 1], dp[i - 1][j]);
                }
            }
        }

        // The bottom-right corner is the length of the LCS
        return dp[m][n];
    }
};
